fsmonitor.c

```diff
@@ -173,27 +189,55 @@ void refresh_fsmonitor(struct index_state *istate)
 	 * should be inclusive to ensure we don't miss potential changes.
 	 */
 	last_update = getnanotime();
-	strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
+	if (hook_version == HOOK_INTERFACE_VERSION1)
+		strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
 
 	/*
-	 * If we have a last update time, call query_fsmonitor for the set of
-	 * changes since that time, else assume everything is possibly dirty
+	 * If we have a last update token, call query_fsmonitor for the set of
+	 * changes since that token, else assume everything is possibly dirty
 	 * and check it all.
 	 */
 	if (istate->fsmonitor_last_update) {
-		query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION,
-			istate->fsmonitor_last_update, &query_result);
+		if (hook_version == -1 || hook_version == HOOK_INTERFACE_VERSION2) {
+			query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION2,
+				istate->fsmonitor_last_update, &query_result);
+
+			if (query_success) {
+				/* First entry will be the last update token */
+				buf = query_result.buf;
+				bol = 0;
+				for (i = 0; i < query_result.len; i++) {
+					if (buf[i] == '\0') {
+						strbuf_addstr(&last_update_token, buf);
+						bol = i + 1;
+						break;
+					}
+				}	
+			}
+		}
+
+		if (hook_version == HOOK_INTERFACE_VERSION1 ||
+		    (hook_version == -1 && !query_success)) {
+			query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION1,
+				istate->fsmonitor_last_update, &query_result);
+
+			if (query_success) {
+				strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
+				buf = query_result.buf;
+				bol = 0;
+				i = 0;
+			}
+		}
+
 		trace_performance_since(last_update, "fsmonitor process '%s'", core_fsmonitor);
 		trace_printf_key(&trace_fsmonitor, "fsmonitor process '%s' returned %s",
 			core_fsmonitor, query_success ? "success" : "failure");
 	}
 
 	/* a fsmonitor process can return '/' to indicate all entries are invalid */
-	if (query_success && query_result.buf[0] != '/') {
+	if (query_success && query_result.buf[bol] != '/') {
 		/* Mark all entries returned by the monitor as dirty */
-		buf = query_result.buf;
-		bol = 0;
-		for (i = 0; i < query_result.len; i++) {
+		for (; i < query_result.len; i++) {
```

Comment:
For v2, don't we start `i` at `bol - 1` by mistake?

In any case, I would prefer changing the code to `for (i = bol; [...])` here, i.e. keep the initial assignment of `i` strictly in this here block.